File: dart_cli/lib/structs/Element.dart

import "../ByteDataWrapper.dart";
import "../DataType.dart";
import "../KeysLookup.dart";
import "../ioClass.dart";
import "JsonArray.dart";
import "JsonFloat.dart";
import "JsonInt.dart";
import "JsonObject.dart";
import "JsonString.dart";

class Element extends JsonIoClass {
  late final DataType dataType;
  late final JsonIoClass? value;

  Element(this.dataType, this.value);

  Element.fromJson(Object? json, KeysLookup keysLookup) {
    (DataType, JsonIoClass?) result;
    if (json is int)
      result = JsonInt.fromJson(json);
    else if (json is double)
      result = JsonFloat.fromJson(json);
    else if (json is bool)
      result = (
        DataType(json ? DataTypeChar.trueBool : DataTypeChar.falseBool),
        null
      );
    else if (json == null)
      result = (DataType(DataTypeChar.nullValue), null);
    else if (json is List)
      result = JsonArray.fromJson(json, keysLookup);
    else if (json is Map)
      result = JsonObject.fromJson(json, keysLookup);
    else if (json is String)
      result = JsonString.fromJson(json);
    else
      throw Exception("Element.fromJson: json is not a valid type: $json");

    dataType = result.$1;
    value = result.$2;
  }

  Element.readBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    dataType = DataType.readBytes(bytes);
    if (!dataType.char.hasJsonValue)
      value = _parseValue(dataType, bytes, keysLookup);
  }

  JsonIoClass _parseValue(DataType dataType, ByteDataWrapper bytes, KeysLookup keysLookup) {
    switch (dataType.char) {
      case DataTypeChar.smallObject:
      case DataTypeChar.bigObject:
      case DataTypeChar.longObject:
        return JsonObject.readBytes(bytes, dataType, keysLookup);
      case DataTypeChar.smallArray:
      case DataTypeChar.bigArray:
      case DataTypeChar.longArray:
        return JsonArray.readBytes(bytes, dataType, keysLookup);
      case DataTypeChar.smallString:
      case DataTypeChar.bigString:
      case DataTypeChar.longString:
        return JsonString.readBytes(bytes, dataType);
      case DataTypeChar.int8:
      case DataTypeChar.uint8:
      case DataTypeChar.int16:
      case DataTypeChar.uint16:
      case DataTypeChar.int32:
      case DataTypeChar.uint32:
      case DataTypeChar.int64:
      case DataTypeChar.uint64:
        return JsonInt.readBytes(bytes, dataType);
      case DataTypeChar.float:
      case DataTypeChar.double:
        return JsonFloat.readBytes(bytes, dataType);
      default:
        throw Exception("Element._parseValue: unexpected data type: ${dataType.char}");
    }
  }

  @override
  Object? toJson() {
    return dataType.char.hasJsonValue
      ? dataType.char.jsonValue
      : value!.toJson();
  }

  @override
  void writeBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    dataType.writeBytes(bytes);
    value?.writeBytes(bytes, keysLookup);    
  }
  
  @override
  int get size {
    var size = dataType.size;
    if (value != null)
      size += value!.size;
    return size;
  }
}

File: python_lib/structs/Element.py
from __future__ import annotations
from DataType import DataType, DataTypeChar
from FileWrapper import BinaryReader, BinaryWriter
from KeysLookup import KeysLookup
from ioClass import JsonIoClass
from structs.JsonArray import JsonArray
from structs.JsonFloat import JsonFloat
from structs.JsonInt import JsonInt
from structs.JsonObject import JsonObject
from structs.JsonString import JsonString


class Element(JsonIoClass):
	dataType: DataType
	value: JsonIoClass|None

	def __init__(self, dataType: DataType, value: JsonIoClass|None):
		self.dataType = dataType
		self.value = value

	@staticmethod
	def fromJson(json: object, keysLookup: KeysLookup) -> Element:
		result: tuple[DataType, JsonIoClass|None]
		if isinstance(json, int):
			result = JsonInt.fromJson(json)
		elif isinstance(json, float):
			result = JsonFloat.fromJson(json)
		elif isinstance(json, bool):
			result = (
				DataType(DataTypeChar.trueBool if json else DataTypeChar.falseBool),
				None
			)
		elif json is None:
			result = (DataType(DataTypeChar.nullValue), None)
		elif isinstance(json, list):
			result = JsonArray.fromJson(json, keysLookup)
		elif isinstance(json, dict):
			result = JsonObject.fromJson(json, keysLookup)
		elif isinstance(json, str):
			result = JsonString.fromJson(json)
		else:
			raise Exception(f"Element.fromJson: json is not a valid type: {json}")
		
		return Element(result[0], result[1])
	
	@staticmethod
	def readBytes(bytes: BinaryReader, keysLookup: KeysLookup) -> Element:
		dataType = DataType.readBytes(bytes)
		value = None
		if not dataType.char.hasJsonValue:
			value = Element._parseValue(dataType, bytes, keysLookup)
		return Element(dataType, value)
	
	@staticmethod
	def _parseValue(dataType: DataType, bytes: BinaryReader, keysLookup: KeysLookup) -> JsonIoClass:
		if dataType.char in [DataTypeChar.smallObject, DataTypeChar.bigObject, DataTypeChar.longObject]:
			return JsonObject.readBytes(bytes, dataType, keysLookup)
		elif dataType.char in [DataTypeChar.smallArray, DataTypeChar.bigArray, DataTypeChar.longArray]:
			return JsonArray.readBytes(bytes, dataType, keysLookup)
		elif dataType.char in [DataTypeChar.smallString, DataTypeChar.bigString, DataTypeChar.longString]:
			return JsonString.readBytes(bytes, dataType)
		elif dataType.char in [DataTypeChar.int8, DataTypeChar.uint8, DataTypeChar.int16, DataTypeChar.uint16, DataTypeChar.int32, DataTypeChar.uint32, DataTypeChar.int64, DataTypeChar.uint64]:
			return JsonInt.readBytes(bytes, dataType)
		elif dataType.char in [DataTypeChar.float, DataTypeChar.double]:
			return JsonFloat.readBytes(bytes, dataType)
		else:
			raise Exception(f"Element._parseValue: unexpected data type: {dataType.char}")
		
	def toJson(self) -> object:
		if self.dataType.char.hasJsonValue:
			return self.dataType.char.jsonValue
		return self.value.toJson()
	
	def writeBytes(self, bytes: BinaryWriter, keysLookup: KeysLookup):
		self.dataType.writeBytes(bytes)
		if self.value is not None:
			self.value.writeBytes(bytes, keysLookup)

	@property
	def size(self) -> int:
		size = self.dataType.size
		if self.value is not None:
			size += self.value.size
		return size
